### PR TITLE
Addition of Debug mode (with compile-option)

### PR DESCRIPTION
**Pull request by twicki**
## Technical Description

Addition of the -fdebug compile fliag that turns off all the passes not needed to generate valid code and putting every statement into it's own mutlistage.

### CODE REVIEW

File: src/dawn/Optimizer/PassMultiStageSplitter.cpp [hunk 1]
```diff
@@ -33,103 +33,157 @@ bool PassMultiStageSplitter::run(
...
-    for(auto multiStageIt = stencil->getMultiStages().begin();
-        multiStageIt != stencil->getMultiStages().end(); ++multiStageIndex) {
-      MultiStage& multiStage = (**multiStageIt);
-
-      splitterIndices.clear();
-      graph.clear();
-
-      // Loop order specified by the user in the vertical region (e.g {k_start, k_end} is forward)
-      auto userSpecifiedLoopOrder = multiStage.getLoopOrder();
+  if(!context->getOptions().Debug) {
```
Comment (cosunae) [inline src/dawn/Optimizer/PassMultiStageSplitter.cpp]: this very large if/elses make the code a bit unreadable. Can we dispatch the code to private methods: ?
```
if(!context->getOptions().Debug) {
   solveMultiStageSplit()
else 
   solveMultiStageSplitDebug();
}
```

File: src/dawn/Optimizer/PassInlining.cpp [hunk 2]
```diff
@@ -465,6 +466,8 @@ bool PassInlining::run(const std::shared_ptr<StencilInstantiation>& stencilInsta
   // Nothing to do ...
   if(strategy_ == IK_None)
     return true;
+  if(stencilInstantiation->getOptimizerContext()->getOptions().Debug)
```
Comment (cosunae) [inline src/dawn/Optimizer/PassInlining.cpp]: this is very hidden here. 
Can we move it to where we push all the passes?  If in debug we should not call the pass, isnt?

File: src/dawn/Optimizer/PassPrintStencilGraph.cpp [hunk 3]
```diff
@@ -25,27 +25,26 @@ PassPrintStencilGraph::PassPrintStencilGraph() : Pass("PassPrintStencilGraph") {
 
 bool PassPrintStencilGraph::run(const std::shared_ptr<StencilInstantiation>& stencilInstantiation) {
   OptimizerContext* context = stencilInstantiation->getOptimizerContext();
-  if(context->getOptions().DumpStencilGraph) {
+  if(!context->getOptions().DumpStencilGraph || context->getOptions().Debug)
+    return true;
```
Comment (cosunae) [inline src/dawn/Optimizer/PassPrintStencilGraph.cpp]: same comment there, I would move this if DEBUG out of the pass. The if (dumpStencilGraph) was already there so we can leave it, but try to not introduce more cases of this type of pattern (with the Debug)

File: src/dawn/Optimizer/PassSSA.cpp [hunk 4]
```diff
@@ -26,63 +26,62 @@ PassSSA::PassSSA() : Pass("PassSSA") {}
 bool PassSSA::run(const std::shared_ptr<StencilInstantiation>& stencilInstantiation) {
   OptimizerContext* context = stencilInstantiation->getOptimizerContext();
 
-  if(context->getOptions().SSA) {
+  if(!context->getOptions().SSA || context->getOptions().Debug)
```
Comment (cosunae) [inline src/dawn/Optimizer/PassSSA.cpp]: same

File: src/dawn/Optimizer/PassSetCaches.cpp [hunk 5]
```diff
@@ -148,6 +148,9 @@ PassSetCaches::PassSetCaches() : Pass("PassSetCaches") {}
 bool PassSetCaches::run(const std::shared_ptr<StencilInstantiation>& instantiation) {
   OptimizerContext* context = instantiation->getOptimizerContext();
 
+  if(context->getOptions().Debug)
```
Comment (cosunae) [inline src/dawn/Optimizer/PassSetCaches.cpp]: same

File: src/dawn/Optimizer/PassSetStageGraph.cpp [hunk 6]
```diff
@@ -61,6 +61,8 @@ PassSetStageGraph::PassSetStageGraph() : Pass("PassSetStageGraph") {
 
 bool PassSetStageGraph::run(const std::shared_ptr<StencilInstantiation>& stencilInstantiation) {
   OptimizerContext* context = stencilInstantiation->getOptimizerContext();
+  if(context->getOptions().Debug)
```
Comment (cosunae) [inline src/dawn/Optimizer/PassSetStageGraph.cpp]: same

File: src/dawn/Optimizer/PassStageReordering.cpp [hunk 7]
```diff
@@ -31,6 +31,9 @@ PassStageReordering::PassStageReordering(ReorderStrategy::ReorderStrategyKind st
 bool PassStageReordering::run(const std::shared_ptr<StencilInstantiation>& stencilInstantiation) {
   OptimizerContext* context = stencilInstantiation->getOptimizerContext();
 
+  if(context->getOptions().Debug)
```
Comment (cosunae) [inline src/dawn/Optimizer/PassStageReordering.cpp]: same for the rest... I ll not comment it further

**Comment by twicki** [comment]
launch jenkins

File: src/dawn/Compiler/DawnCompiler.cpp [hunk 8]
```diff
@@ -172,6 +186,7 @@ std::unique_ptr<OptimizerContext> DawnCompiler::runOptimizer(std::shared_ptr<SIR
                    << "` ...";
     if(!passManager.runAllPassesOnStecilInstantiation(instantiation))
       return nullptr;
+//    instantiation->dump();
```
Comment (cosunae) [inline src/dawn/Compiler/DawnCompiler.cpp]: remove. 
btw,we should have a compiler flag that just runs this dump

File: src/dawn/Optimizer/OptimizerContext.h [hunk 9]
```diff
@@ -80,6 +80,31 @@ class OptimizerContext : NonCopyable {
   /// @brief Get the hardware configuration
   const HardwareConfig& getHardwareConfiguration() const { return hardwareConfiguration_; }
   HardwareConfig& getHardwareConfiguration() { return hardwareConfiguration_; }
+
+  /// @brief Create a new pass at the end of the pass list
+  template <class T, typename... Args>
+  void checkAndPushBack(Args&&... args) {
+    std::unique_ptr<T> pass = make_unique<T>(std::forward<Args>(args)...);
+    if(compareOptionsToPassFlags<T>(pass)) {
+      passManager_.getPasses().push_back(std::move(pass));
```
Comment (cosunae) [inline src/dawn/Optimizer/OptimizerContext.h]: neat

File: src/dawn/Optimizer/PassDataLocalityMetric.cpp [hunk 10]
```diff
@@ -310,6 +310,9 @@ bool PassDataLocalityMetric::run(
     const std::shared_ptr<StencilInstantiation>& stencilInstantiation) {
   OptimizerContext* context = stencilInstantiation->getOptimizerContext();
 
+  if(context->getOptions().Debug)
```
Comment (cosunae) [inline src/dawn/Optimizer/PassDataLocalityMetric.cpp]: this should be gone, isnt?

File: src/dawn/Optimizer/PassMultiStageSplitter.cpp [hunk 11]
```diff
@@ -27,101 +27,169 @@
...
+      graph.toDot(format("stmt_vd_m%i_%02i.dot", multiStageIndex, numSplit));
 
-          // Check for read-before-write conflicts in the loop order and counter loop order.
-          // Conflicts in the loop order will assure us that the multi-stage can't be
-          // parallel. A conflict in the counter loop order is more severe and needs the current
-          // multistage be splitted!
-          auto conflict = hasVerticalReadBeforeWriteConflict(&graph, userSpecifiedLoopOrder);
+    if(!splitterIndices.empty()) {
+      auto newMultiStages = multiStage.split(splitterIndices, curLoopOrder);
+      multiStageIt = stencil->getMultiStages().erase(multiStageIt);
```
Comment (cosunae) [inline src/dawn/Optimizer/PassMultiStageSplitter.cpp]: this code is also common. Is it possible to move this out of the lambda, and have the lambda only inside the loop over stages?

File: src/dawn/Optimizer/PassMultiStageSplitter.cpp [hunk 12]
```diff
@@ -27,101 +27,169 @@
...
+PassMultiStageSplitter::PassMultiStageSplitter(MulitStageSplittingStrategy strategy)
+    : Pass("PassMultiStageSplitter"), strategy_(strategy) {
+  isDebug_ = true;
+}
+namespace {
+std::function<void(std::list<std::shared_ptr<MultiStage>>::iterator&, OptimizerContext*, int&,
+                   DependencyGraphAccesses&, std::shared_ptr<Stencil>, int&, std::string&)>
+setOptimizedLoopContent(const std::shared_ptr<StencilInstantiation>& stencilInstantiation) {
+  return [&](std::list<std::shared_ptr<MultiStage>>::iterator& multiStageIt,
+             OptimizerContext* context, int& numSplit, DependencyGraphAccesses& graph,
```
Comment (cosunae) [inline src/dawn/Optimizer/PassMultiStageSplitter.cpp]: context is not really needed, we can simply pass the boolean flags used, 
doDump...

**Pull request merged**